.travis.yml

```diff
@@ -3,35 +3,48 @@ sudo: required
 service:
 - docker
 - docker-compose
-jobs:
+
+install:
+- wget https://releases.hashicorp.com/terraform/"$TF_VERSION"/terraform_"$TF_VERSION"_linux_amd64.zip 
+- unzip terraform_"$TF_VERSION"_linux_amd64.zip 
+- sudo mv terraform /usr/local/bin/ 
+- rm terraform_"$TF_VERSION"_linux_amd64.zip
+
+jobs:  
   include:
-    - stage: build and test
+    - stage: Test
+      if: branch != master
       script:
+      - cd terraform_env/dev
+      - terraform init
+      - terraform plan -var "prefix=test" -var "location=$LOCATION" -var "resource_group_name=$RESOURCE_GROUP_NAME"
+      - terraform apply -var "prefix=test" -var "location=$LOCATION" -var "resource_group_name=$RESOURCE_GROUP_NAME" -auto-approve
+      - export MONGO_CONNECTION_STRING="$(terraform output -raw test_cosmosdb_connection_string)"
+      - cd ../..
       - docker-compose -f docker-compose-dev.yml up -d --build
-      - docker run --env-file .env.test --entrypoint poetry
-        todo-app-test:latest run pytest tests
-      - docker run -e MONGO_CONNECTION_STRING -e MONGODB_COLLECTIONNAME -e SECRET_KEY -e GITHUB_CLIENT_ID -e GITHUB_CLIENT_SECRET  --entrypoint poetry
-        todo-app-test:latest run pytest tests_e2e
-
-
-before_deploy:
-- docker build --target production --tag $dockerUsername/todo-app-prod:latest .
-- echo "$dockerPassword" | docker login -u "$dockerUsername" --password-stdin
-- docker push $dockerUsername/todo-app-prod:latest
-
-deploy:
-  provider: script
-  script: bash ./deploy.sh
-  on:
-    branch: master
-  
-  provider: script
-  script: bash ./webhook.sh
-  on:
-    branch: master
-
-after_success:
-- echo "All done!"
+      - docker run --env-file .env.test --entrypoint poetry todo-app-test:latest run pytest tests
+      - docker run -e MONGO_CONNECTION_STRING -e MONGODB_COLLECTION_NAME -e SECRET_KEY -e GITHUB_CLIENT_ID -e GITHUB_CLIENT_SECRET --entrypoint poetry todo-app-test:latest run pytest tests_e2e
+      - docker stop $(docker ps --format "{{.ID}}")
+      - cd ./terraform_env/dev
+      - terraform destroy -var "prefix=test" -var "location=$LOCATION" -var "resource_group_name=$RESOURCE_GROUP_NAME" -auto-approve
+    - stage: Docker Push
+      if: branch = master
+      script:
+      - docker build --target production --tag $DOCKER_USERNAME/todo-app-prod:latest .
+      - echo "$DOCKER_PASSWORD" | docker login -u "$DOCKER_USERNAME" --password-stdin
+      - docker push $DOCKER_USERNAME/todo-app-prod:latest
+    - stage: Publish Production
+      if: branch = master
+      script:
+      - cd terraform_env/prod
+      - terraform init
+      - terraform plan -var "prefix=prod" -var "location=$LOCATION" -var "RESOURCE_GROUP_NAME=$RESOURCE_GROUP_NAME" -var "DOCKER_REGISTRY_SERVER_USERNAME=$DOCKER_USERNAME" -var "DOCKER_REGISTRY_SERVER_PASSWORD=$DOCKER_PASSWORD" -var "GITHUB_CLIENT_ID=$PROD_GITHUB_CLIENT_ID" -var "GITHUB_CLIENT_SECRET=$PROD_GITHUB_CLIENT_SECRET" -var "MONGODB_COLLECTION_NAME=$MONGODB_COLLECTION_NAME" -var "SECRET_KEY=$SECRET_KEY"
```

Comment:
Is there any reason you run a plan in an automated pipeline that then runs a separate apply after? `terraform apply` will run a fresh plan anyway unless you provide the output from an existing one